Track workflow step input definitions in our model.

jmchilton opened the pull request:
We don't track workflow step inputs in any formal way in our model currently. This has resulted in some current hacks and prevents future enhancements. This commit splits WorkflowStepConnection into two models WorkflowStepInput and WorkflowStepConnection - normalizing the previous table workflow_step_connection on input step and input name.

In terms of current hacks forced on it by restricting all of tool state to be confined to a big JSON blob in the database - we have problems distinguishing keys and values when walking tool state. As we store more and more JSON blobs inside of the giant tool state blob - the worse this problem gets. Take for instance checking for runtime parameters or the rules parameter values - these both use JSON blobs that aren't simple values, so it is hard to tell looking at the tool state blob in the database or the workflow export to tell what is a key or what is a value. Tracking state as normalized inputs with default values and explicit attributes runtime values should allow much more percise state definition and construction.

This variant of the models would also potentially allow defining runtime values with non-tool default values (so default values defined for the workflow but still explicitly settable at runtime). The combinations of overriding defaults and defining runtime values were not representable before.

In terms of future enhancements, there is a lot we cannot track with the current models - such as map/reduce options for collection operations (https://github.com/galaxyproject/galaxy/issues/4623#issuecomment-389544980). This should enable a lot of that. Obviously there are a lot of attributes defined here that are not yet utilized, but I'm using most (all?) of them downstream in the CWL branch. I'd rather populate this table fully realized and fill in the implementation around it as work continues to stream in from the CWL branch - to keep things simple and avoid extra database migrations. But I understand if this feels like speculative complexity we want to avoid despite the implementation being readily available for inspection downstream.

jmchilton committed:
Track workflow step input definitions in our model.

We don't track workflow step inputs in any formal way in our model currently. This has resulted in some current hacks and prevents future enhancements. This commit splits WorkflowStepConnection into two models WorkflowStepInput and WorkflowStepConnection - normalizing the previous table workflow_step_connection on input step and input name.

In terms of current hacks forced on it by restricting all of tool state to be confined to a big JSON blob in the database - we have problems distinguishing keys and values when walking tool state. As we store more and more JSON blobs inside of the giant tool state blob - the worse this problem gets. Take for instance checking for runtime parameters or the rules parameter values - these both use JSON blobs that aren't simple values, so it is hard to tell looking at the tool state blob in the database or the workflow export to tell what is a key or what is a value. Tracking state as normalized inputs with default values and explicit attributes runtime values should allow much more percise state definition and construction.

This variant of the models would also potentially allow defining runtime values with non-tool default values (so default values defined for the workflow but still explicitly settable at runtime). The combinations of overriding defaults and defining runtime values were not representable before.

In terms of future enhancements, there is a lot we cannot track with the current models - such as map/reduce options for collection operations (https://github.com/galaxyproject/galaxy/issues/4623#issuecomment-389544980). This should enable a lot of that. Obviously there are a lot of attributes defined here that are not yet utilized, but I'm using most (all?) of them downstream in the CWL branch. I'd rather populate this table fully realized and fill in the implementation around it as work continues to stream in from the CWL branch - to keep things simple and avoid extra database migrations. But I understand if this feels like speculative complexity we want to avoid despite the implementation being readily available for inspection downstream.

bgruening commented:
> But I understand if this feels like speculative complexity we want to avoid despite the implementation being readily available for inspection downstream.

I can not judge this and this PR feels like a good first step. Thanks John!
In regards to the json blobs I wanted to point out that we, in addition, also might want to consider to move over to (JsonTypes)[https://docs.sqlalchemy.org/en/latest/core/type_basics.html#sqlalchemy.types.JSON]. This feature is long enough in all our supported RDMS and could make queries faster.

nsoranzo commented on lib/galaxy/dependencies/pipfiles/default/pinned-requirements.txt:
```diff
@@ -56,7 +56,7 @@ functools32==3.2.3.post2 ; python_version == '2.7'
 futures==3.2.0 ; python_version == '2.6' or python_version == '2.7'
 galaxy-sequence-utils==1.1.3
-gxformat2==0.7.1
+gxformat2
```
```suggestion
gxformat2==0.8.0
```

nsoranzo commented on lib/galaxy/model/mapping.py:
```diff
@@ -908,6 +908,21 @@
+    "workflow_step_input", metadata,
+    Column("id", Integer, primary_key=True),
+    Column("workflow_step_id", Integer, ForeignKey("workflow_step.id"), index=True),
+    Column("name", Unicode(255)),
```
```suggestion
    Column("name", TEXT),
```

nsoranzo commented on lib/galaxy/model/mapping.py:
```diff
@@ -908,6 +908,21 @@
+    Column("value_from", JSONType),
+    Column("value_from_type", TEXT),
+    Column("default_value", JSONType),
+    Column("default_value_set", Boolean),
```
```suggestion
    Column("default_value_set", Boolean, default=False),
```

nsoranzo replied:
Can we get rid of `default_value_set` column and just check that `default_value is not None` ?

jmchilton replied:
No - because default_value can be null for certain datatypes?

nsoranzo commented on lib/galaxy/model/mapping.py:
```diff
@@ -908,6 +908,21 @@
+    Column("value_from_type", TEXT),
+    Column("default_value", JSONType),
+    Column("default_value_set", Boolean),
+    Column("runtime_value", Boolean))
```
```suggestion
    Column("runtime_value", Boolean, default=False))
```

nsoranzo commented:
The failing `selenium_tests.test_workflow_run.WorkflowRunTestCase.test_subworkflow_runtime_parameters` selenium test seems legit.

nsoranzo commented on lib/galaxy/webapps/tool_shed/model/__init__.py:
```diff
@@ -479,10 +479,39 @@ def __init__(self):
+        connections = []
+        for step_input in self.inputs:
+            for connection in step_input.connections:
+                connections.append(connection)
```
You can replace the above 4 lines with:

```python
        connections = [_ for step_input in self.inputs for _ in step_input.connections]
```

nsoranzo commented on lib/galaxy/model/__init__.py:
```diff
@@ -4071,13 +4071,49 @@ def __init__(self):
+        connections = []
+        for step_input in self.inputs:
+            for connection in step_input.connections:
+                connections.append(connection)
```
You can replace the above 4 lines with:

```python
        connections = [_ for step_input in self.inputs for _ in step_input.connections]
```

nsoranzo commented on lib/galaxy/webapps/tool_shed/model/__init__.py:
```diff
@@ -479,10 +479,39 @@ def __init__(self):
         self.config = None
         self.label = None
 
+        self._input_connections_by_name = None
```
Unused?

nsoranzo commented on lib/galaxy/model/__init__.py:
```diff
@@ -4071,13 +4071,49 @@ def __init__(self):
+        if step_input is None:
+            step_input = WorkflowStepInput(self)
+            step_input.name = input_name
+            # self.inputs.append(step_input)
```
Remove?

nsoranzo commented on lib/galaxy/workflow/modules.py:
```diff
@@ -1042,14 +1057,14 @@ def get_runtime_state(self):
         return self.get_inputs()
 
-    def compute_runtime_state(self, trans, step_updates=None):
+    def compute_runtime_state(self, trans, step, step_updates=None):
```
```suggestion
    def compute_runtime_state(self, trans, step=None, step_updates=None):
```

nsoranzo commented on lib/galaxy/workflow/modules.py:
```diff
@@ -197,7 +197,7 @@ def get_runtime_inputs(self, **kwds):
         return {}
 
-    def compute_runtime_state(self, trans, step_updates=None):
+    def compute_runtime_state(self, trans, step=None, step_updates=None):
```
Can you document the new `step` parameter, please?

jmchilton replied:
Done below.

nsoranzo commented on lib/galaxy/managers/workflows.py:
```diff
@@ -998,6 +1006,18 @@ def __module_from_dict(self, trans, steps, steps_by_external_id, step_dict, **kw
+                    step_input.default_value = default
+                    step_input.default_value_set = True
+
+                step.get_or_add_input(input_name)
```
I don't understand why this line is needed.

nsoranzo commented on lib/galaxy/model/__init__.py:
```diff
@@ -4178,6 +4213,35 @@ def log_str(self):
+        copied_step_input.connections = copy_list(self.connections)
+        return copied_step_input
+
+    def log_str(self):
```
Unused?

nsoranzo commented on lib/galaxy/model/__init__.py:
```diff
@@ -4189,18 +4253,29 @@ class WorkflowStepConnection(RepresentById):
         # TODO: handle subworkflow ids...
         copied_connection = WorkflowStepConnection()
         copied_connection.output_name = self.output_name
-        copied_connection.input_name = self.input_name
```
This method seems broken in the current state, it copies just the `output_name`.

jmchilton replied:
We don't yet have the new ids for the new step inputs yet, they get re-assigned later on when copying workflow steps.

nsoranzo commented on lib/galaxy/model/migrate/versions/0145_add_workflow_step_input.py:
```diff
@@ -0,0 +1,105 @@
+    LegacyWorkflowStepConnection_table = Table("workflow_step_connection", metadata, autoload=True)
+    for index in LegacyWorkflowStepConnection_table.indexes:
+        index.drop()
+    LegacyWorkflowStepConnection_table.rename("workflow_step_connection_premigrate144")
```
Should this be updated to the current migration number (also in other places below)?

```suggestion
    LegacyWorkflowStepConnection_table.rename("workflow_step_connection_premigrate145")
```

jmchilton replied:
I'll fix this, but only if you merge before we leave tonight 😄.

jmchilton commented on lib/galaxy/model/__init__.py:
```diff
@@ -4174,8 +4205,29 @@ def copy_to(self, copied_step, step_mapping):
                           action_arguments=orig_pja.action_arguments)
         copied_step.workflow_outputs = copy_list(self.workflow_outputs, copied_step)
 
-    def log_str(self):
```
Removed the wrong thing.

jmchilton commented:
Indeed the Selenium failure seems legit:

```
Traceback (most recent call last):
  File "/Users/john/workspace/galaxy/lib/galaxy/web/framework/decorators.py", line 283, in decorator
    rval = func(self, trans, *args, **kwargs)
  File "/Users/john/workspace/galaxy/lib/galaxy/webapps/galaxy/api/workflows.py", line 414, in workflow_dict
    ret_dict = self.workflow_contents_manager.workflow_to_dict(trans, stored_workflow, style=style, version=version)
  File "/Users/john/workspace/galaxy/lib/galaxy/managers/workflows.py", line 437, in workflow_to_dict
    wf_dict = self._workflow_to_dict_run(trans, stored, workflow=workflow)
  File "/Users/john/workspace/galaxy/lib/galaxy/managers/workflows.py", line 468, in _workflow_to_dict_run
    module_injector.inject(step, steps=workflow.steps, exact_tools=False)
  File "/Users/john/workspace/galaxy/lib/galaxy/workflow/modules.py", line 1425, in inject
    state, step_errors = module.compute_runtime_state(self.trans, step, step_args)
  File "/Users/john/workspace/galaxy/lib/galaxy/workflow/modules.py", line 230, in compute_runtime_state
    visit_input_values(self.get_runtime_inputs(), state.inputs, update_value, no_replacement_value=NO_REPLACEMENT)
  File "/Users/john/workspace/galaxy/lib/galaxy/tools/parameters/__init__.py", line 167, in visit_input_values
    callback_helper(input, input_values, name_prefix, label_prefix, parent_prefix=parent_prefix, context=context)
  File "/Users/john/workspace/galaxy/lib/galaxy/tools/parameters/__init__.py", line 121, in callback_helper
    'value'             : input_values.get(input.name),
AttributeError: 'InputProxy' object has no attribute 'name'
```

nsoranzo commented on lib/galaxy/model/mapping.py:
```diff
@@ -908,6 +908,21 @@
+    Column("value_from_type", TEXT),
+    Column("default_value", JSONType),
+    Column("default_value_set", Boolean, default=False),
+    Column("runtime_value", Boolean))
```
I think this got lost in the rebase?

```suggestion
    Column("runtime_value", Boolean, default=False))
```

nsoranzo committed:
Update lib/galaxy/model/mapping.py

Co-Authored-By: jmchilton <jmchilton@gmail.com>

nsoranzo commented on lib/galaxy/model/mapping.py:
```diff
@@ -908,6 +908,21 @@
+    Column("value_from_type", TEXT),
+    Column("default_value", JSONType),
+    Column("default_value_set", Boolean, default=False),
+    Column("runtime_value", Boolean, default=False))
```
What about enforcing uniqueness of an input `name` for a workflow step with:
```
    UniqueConstraint("workflow_step_id", "name", name='uix_1')
```

jmchilton replied:
This is a great idea!

nsoranzo commented on lib/galaxy/model/migrate/versions/0145_add_workflow_step_input.py:
```diff
@@ -0,0 +1,105 @@
+        Column("value_from_type", TEXT),
+        Column("default_value", JSONType),
+        Column("default_value_set", Boolean, default=False),
+        Column("runtime_value", Boolean, default=False),
```
`UniqueConstraint("workflow_step_id", "name", name='uix_1'),`

nsoranzo commented on lib/galaxy/model/migrate/versions/0145_add_workflow_step_input.py:
```diff
@@ -0,0 +1,105 @@
+    insert_step_connections_cmd = \
+        "INSERT INTO workflow_step_connection (output_step_id, input_step_input_id, output_name, input_subworkflow_step_id) " + \
+        "SELECT wsc.output_step_id, wsi.id, wsc.output_name, wsc.input_subworkflow_step_id " + \
+        "FROM workflow_step_connection_premigrate145 as wsc left outer join workflow_step_input as wsi on wsc.input_step_id = wsi.workflow_step_id and wsc.input_name = wsi.name ORDER BY wsc.id"
```
```suggestion
        "FROM workflow_step_connection_premigrate145 AS wsc LEFT OUTER JOIN workflow_step_input AS wsi ON wsc.input_step_id = wsi.workflow_step_id AND wsc.input_name = wsi.name ORDER BY wsc.id"
```

nsoranzo commented on lib/galaxy/model/migrate/versions/0145_add_workflow_step_input.py:
```diff
@@ -0,0 +1,105 @@
+
+    # Drop new workflow invocation step and job association table and restore legacy data.
+    LegacyWorkflowStepConnection_table = Table("workflow_step_connection_premigrate145", metadata, autoload=True)
+    LegacyWorkflowStepConnection_table.rename("workflow_step_connection")
```
This will lose any update to the `workflow_step_connection` table after the upgrade, is it worth to try to instead to recreate the table using the old schema but with the present content? I can work on that, if you agree.

jmchilton replied:
I don't think it is worth the effort, but I wouldn't stop you from working on it if you wanted to.

nsoranzo replied:
OK, thanks, I'll give it a try and push to your branch if successful.

nsoranzo committed:
Update lib/galaxy/model/migrate/versions/0145_add_workflow_step_input.py

Co-Authored-By: jmchilton <jmchilton@gmail.com>

nsoranzo commented on lib/galaxy/model/migrate/versions/0145_add_workflow_step_input.py:
```diff
@@ -0,0 +1,105 @@
+
+    insert_step_inputs_cmd = \
+        "INSERT INTO workflow_step_input (workflow_step_id, name) " + \
+        "SELECT id, input_name FROM workflow_step_connection_premigrate145"
```
@jmchilton I think I found a major issue with the migration, can you double-check that the following is correct?
```suggestion
        "SELECT input_step_id, input_name FROM workflow_step_connection_premigrate145"
```

If confirmed, I'll add it to my forthcoming commit.

nsoranzo committed:
Update lib/galaxy/model/migrate/versions/0145_add_workflow_step_input.py

Co-Authored-By: jmchilton <jmchilton@gmail.com>

nsoranzo committed:
Preserve updated content in migration 145 downgrade

Also add UNIQUE constraint to `("workflow_step_id", "name")` to the
new `workflow_step_input` table.

mvdbeek commented:
Hmm, the migration fails on my dev instance:
```
Activating virtualenv at .venv3
WARNING: Your Python version is: 3.6.5
Galaxy is currently supported on Python 2.7 only, support for Python >= 3.4
is still in beta stage. Since you are using a virtual environment, we assume
you are developing or testing Galaxy under Python 3. If not,
install a supported Python version. If a supported version is already
installed but is not your default, https://galaxyproject.org/admin/python/
contains instructions on how to force Galaxy to use a different version.
144 -> 145...

Migration script for workflow step input table.

Traceback (most recent call last):
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1193, in _execute_context
    context)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 509, in do_execute
    cursor.execute(statement, parameters)
psycopg2.IntegrityError: duplicate key value violates unique constraint "workflow_step_input_workflow_step_id_name_key"
DETAIL:  Key (workflow_step_id, name)=(45, multibam_conditional|bamfiles) already exists.


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "./scripts/manage_db.py", line 24, in <module>
    invoke_migrate_main()
  File "./scripts/manage_db.py", line 20, in invoke_migrate_main
    main(repository=repo, url=db_url)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/migrate/versioning/shell.py", line 209, in main
    ret = command_func(**kwargs)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/migrate/versioning/api.py", line 186, in upgrade
    return _migrate(url, repository, version, upgrade=True, err=err, **opts)
  File "<decorator-gen-15>", line 2, in _migrate
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/migrate/versioning/util/__init__.py", line 167, in with_engine
    return f(*a, **kw)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/migrate/versioning/api.py", line 366, in _migrate
    schema.runchange(ver, change, changeset.step)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/migrate/versioning/schema.py", line 93, in runchange
    change.run(self.engine, step)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/migrate/versioning/script/py.py", line 148, in run
    script_func(engine)
  File "lib/galaxy/model/migrate/versions/0145_add_workflow_step_input.py", line 69, in upgrade
    migrate_engine.execute(insert_step_inputs_cmd)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 2075, in execute
    return connection.execute(statement, *multiparams, **params)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 942, in execute
    return self._execute_text(object, multiparams, params)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1104, in _execute_text
    statement, parameters
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1200, in _execute_context
    context)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1413, in _handle_dbapi_exception
    exc_info
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 265, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 248, in reraise
    raise value.with_traceback(tb)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1193, in _execute_context
    context)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 509, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.IntegrityError: (psycopg2.IntegrityError) duplicate key value violates unique constraint "workflow_step_input_workflow_step_id_name_key"
DETAIL:  Key (workflow_step_id, name)=(45, multibam_conditional|bamfiles) already exists.
 [SQL: 'INSERT INTO workflow_step_input (workflow_step_id, name) SELECT input_step_id, input_name FROM workflow_step_connection_preupgrade145'] (Background on this error at: http://sqlalche.me/e/gkpj)
```

nsoranzo commented:
@mvdbeek Can you try something like the following (not tested):
```
diff --git a/lib/galaxy/model/migrate/versions/0145_add_workflow_step_input.py b/lib/galaxy/model/migrate/versions/0145_add_workflow_step_input.py
index d648d4b..cbbd2b9 100644
--- a/lib/galaxy/model/migrate/versions/0145_add_workflow_step_input.py
+++ b/lib/galaxy/model/migrate/versions/0145_add_workflow_step_input.py
@@ -65,7 +65,7 @@ def upgrade(migrate_engine):
 
     insert_step_inputs_cmd = \
         "INSERT INTO workflow_step_input (workflow_step_id, name) " + \
-        "SELECT input_step_id, input_name FROM workflow_step_connection_preupgrade145"
+        "SELECT DISTINCT input_step_id, input_name FROM workflow_step_connection_preupgrade145"
     migrate_engine.execute(insert_step_inputs_cmd)
 
     insert_step_connections_cmd = \
```

mvdbeek commented:
slightly different error now:
```
144 -> 145...

Migration script for workflow step input table.

Traceback (most recent call last):
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1193, in _execute_context
    context)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 509, in do_execute
    cursor.execute(statement, parameters)
psycopg2.ProgrammingError: relation "workflow_step_connection_preupgrade145" already exists


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "./scripts/manage_db.py", line 24, in <module>
    invoke_migrate_main()
  File "./scripts/manage_db.py", line 20, in invoke_migrate_main
    main(repository=repo, url=db_url)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/migrate/versioning/shell.py", line 209, in main
    ret = command_func(**kwargs)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/migrate/versioning/api.py", line 186, in upgrade
    return _migrate(url, repository, version, upgrade=True, err=err, **opts)
  File "<decorator-gen-15>", line 2, in _migrate
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/migrate/versioning/util/__init__.py", line 167, in with_engine
    return f(*a, **kw)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/migrate/versioning/api.py", line 366, in _migrate
    schema.runchange(ver, change, changeset.step)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/migrate/versioning/schema.py", line 93, in runchange
    change.run(self.engine, step)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/migrate/versioning/script/py.py", line 148, in run
    script_func(engine)
  File "lib/galaxy/model/migrate/versions/0145_add_workflow_step_input.py", line 49, in upgrade
    OldWorkflowStepConnection_table.rename("workflow_step_connection_preupgrade145")
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/migrate/changeset/schema.py", line 507, in rename
    run_single_visitor(engine, visitorcallable, self, connection, **kwargs)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/migrate/changeset/databases/visitor.py", line 85, in run_single_visitor
    fn(element, **kwargs)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/migrate/changeset/ansisql.py", line 164, in visit_table
    self.execute()
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/migrate/changeset/ansisql.py", line 44, in execute
    return self.connection.execute(self.buffer.getvalue())
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 942, in execute
    return self._execute_text(object, multiparams, params)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1104, in _execute_text
    statement, parameters
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1200, in _execute_context
    context)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1413, in _handle_dbapi_exception
    exc_info
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 265, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 248, in reraise
    raise value.with_traceback(tb)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1193, in _execute_context
    context)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 509, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.ProgrammingError: (psycopg2.ProgrammingError) relation "workflow_step_connection_preupgrade145" already exists
 [SQL: '\nALTER TABLE workflow_step_connection RENAME TO workflow_step_connection_preupgrade145'] (Background on this error at: http://sqlalche.me/e/f405)
```

mvdbeek commented:
Might just need to drop `workflow_step_connection_preupgrade145 ` ?

nsoranzo commented:
Not drop it, rename it back to `workflow_step_connection` !
